vm/block.go

```diff
@@ -13,313 +13,78 @@ import (
 	ctx, span := vm.Tracer().Start(ctx, "chain.ParseBlock")
 	defer span.End()
 
-	blk, err := UnmarshalBlock(source, vm)
+	blk, err := vm.chain.ParseBlock(ctx, source)
 	if err != nil {
 		return nil, err
 	}
 	// Not guaranteed that a parsed block is verified
-	return ParseStatefulBlock(ctx, blk, source, accepted, vm)
-}
-
-// populateTxs is only called on blocks we did not build
-func (b *StatefulBlock) populateTxs(ctx context.Context) error {
-	ctx, span := b.vm.Tracer().Start(ctx, "StatefulBlock.populateTxs")
-	defer span.End()
-
-	// Setup signature verification job
-	_, sigVerifySpan := b.vm.Tracer().Start(ctx, "StatefulBlock.verifySignatures") //nolint:spancheck
-	job, err := b.vm.AuthVerifiers().NewJob(len(b.Txs))
-	if err != nil {
-		return err //nolint:spancheck
-	}
-	b.sigJob = job
-	batchVerifier := NewAuthBatch(b.vm, b.sigJob, b.authCounts)
-
-	// Make sure to always call [Done], otherwise we will block all future [Workers]
-	defer func() {
-		// BatchVerifier is given the responsibility to call [b.sigJob.Done()] because it may add things
-		// to the work queue async and that may not have completed by this point.
-		go batchVerifier.Done(func() { sigVerifySpan.End() })
-	}()
-
-	// Confirm no transaction duplicates and setup
-	// AWM processing
-	b.txsSet = set.NewSet[ids.ID](len(b.Txs))
-	for _, tx := range b.Txs {
-		// Ensure there are no duplicate transactions
-		if b.txsSet.Contains(tx.ID()) {
-			return ErrDuplicateTx
-		}
-		b.txsSet.Add(tx.ID())
-
-		// Verify signature async
-		if b.vm.GetVerifyAuth() {
-			unsignedTxBytes, err := tx.UnsignedBytes()
-			if err != nil {
-				return err
-			}
-			batchVerifier.Add(unsignedTxBytes, tx.Auth)
-		}
-	}
-	return nil
+	return ParseStatefulBlock(ctx, blk, accepted, vm)
 }
 
 func ParseStatefulBlock(
 	ctx context.Context,
-	blk *StatelessBlock,
-	source []byte,
+	blk *chain.ExecutionBlock,
 	accepted bool,
-	vm VM,
+	vm *VM,
 ) (*StatefulBlock, error) {
 	ctx, span := vm.Tracer().Start(ctx, "chain.ParseStatefulBlock")
 	defer span.End()
 
 	// Perform basic correctness checks before doing any expensive work
-	if blk.Tmstmp > time.Now().Add(FutureBound).UnixMilli() {
-		return nil, ErrTimestampTooLate
+	if blk.Tmstmp > time.Now().Add(chain.FutureBound).UnixMilli() {
+		return nil, chain.ErrTimestampTooLate
 	}
 
-	if len(source) == 0 {
-		nsource, err := blk.Marshal()
-		if err != nil {
-			return nil, err
-		}
-		source = nsource
-	}
 	b := &StatefulBlock{
-		StatelessBlock: blk,
+		ExecutionBlock: blk,
 		t:              time.UnixMilli(blk.Tmstmp),
-		bytes:          source,
 		accepted:       accepted,
 		vm:             vm,
-		executor:       NewExecutor(vm),
-		id:             utils.ToID(source),
-	}
-
-	// If we are parsing an older block, it will not be re-executed and should
-	// not be tracked as a parsed block
-	lastAccepted := b.vm.LastAcceptedBlock()
-	if lastAccepted == nil || b.Hght <= lastAccepted.Hght { // nil when parsing genesis
-		return b, nil
```

Comment:
Should add this optimization back into chain. Note: doing this early can cause us to do unnecessary signature verification of any block that we are willing to parse